### PR TITLE
fix compilation error in tensor.h

### LINKED ISSUE #8387: Current master has compilation errors related to constexpr in tensor.h
When trying to compile on the palmetto cluster in Clemson I have come across the following errors when trying to compile the current master version: 

_required from here
/home/tcleven/deal-git/dealii/include/deal.II/base/tensor.h:1933:5: error: uninitialized variable ‘result’ in ‘constexpr’ function
     result;
     ^~~~~~

This same error occurs in many places. 

It should be noted:
DEAL_II_WITH_CXX14 = ON
DEAL_II_WITH_CXX17 = OFF 
I'm attaching the detailed.log. Let me know if you need any other information from me.
[detailed.log](https://github.com/dealii/dealii/files/3398921/detailed.log)

### PR DESCRIPTION
**Pull request by tcclevenger**
Fixes #8387. As discussed in https://github.com/dealii/dealii/issues/8387. I encountered an error when compiling the latest master version of dealii. Fix was suggested by @masterleinad.

### CODE REVIEW

**Comment by masterleinad** [comment]
Would you mind modifying the remaining places where we are using `tensor_type` as well?

**Comment by tcclevenger** [comment]
> Would you mind modifying the remaining places where we are using `tensor_type` as well?

No problem. I have gotten all of them in tensor.h

File: include/deal.II/base/tensor.h [hunk 1]
```diff
@@ -2138,7 +2138,7 @@ DEAL_II_CONSTEXPR inline DEAL_II_ALWAYS_INLINE
   scalar_product(const Tensor<rank, dim, Number> &     left,
                  const Tensor<rank, dim, OtherNumber> &right)
 {
-  typename ProductType<Number, OtherNumber>::type result;
+  typename ProductType<Number, OtherNumber>::type result{};
```
Comment (masterleinad) [inline include/deal.II/base/tensor.h]: I am surprised that you had to modify this line as well. Did you get similar errors?

Comment (tcclevenger) [reply include/deal.II/base/tensor.h]: Yes, this is the only other line I got an error on.

**Pull request merged**